pandas/core/arrays/string_.py

```diff
@@ -0,0 +1,217 @@
+    ValueError: Must provide strings
+    """
+
+    # undo the PandasArray hack
+    _typ = "extension"
+
+    def __init__(self, values, copy=False):
+        super().__init__(values, copy=copy)
+        self._dtype = StringDtype()
+        self._validate()
+
+    def _validate(self):
+        """Validate that we only store NA or strings."""
+        if len(self._ndarray) and not lib.is_string_array(self._ndarray, skipna=True):
+            raise ValueError("Must provide strings")
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        if dtype:
+            assert dtype == "string"
+        result = super()._from_sequence(scalars, dtype=object, copy=copy)
+        # convert None to np.nan
+        # TODO: it would be nice to do this in _validate / lib.is_string_array
+        # We are already doing a scan over the values there.
+        result[result.isna()] = np.nan
+        return result
+
+    @classmethod
+    def _from_sequence_of_strings(cls, strings, dtype=None, copy=False):
+        return cls._from_sequence(strings, dtype=dtype, copy=copy)
+
+    def __setitem__(self, key, value):
+        value = extract_array(value, extract_numpy=True)
+        if isinstance(value, type(self)):
+            value = value._ndarray
+        scalar_key = lib.is_scalar(key)
+        scalar_value = lib.is_scalar(value)
+        if scalar_key and not scalar_value:
+            raise ValueError("setting an array element with a sequence.")
+
+        # validate new items
+        if scalar_value:
+            if scalar_value is None:
+                value = np.nan
+            elif not (isinstance(value, str) or np.isnan(value)):
+                raise ValueError(
+                    "Cannot set value '{}' into a StringArray.".format(value)
+                )
+        else:
+            if not is_array_like(value):
+                value = np.asarray(value, dtype=object)
+            if len(value) and not lib.is_string_array(value, skipna=True):
+                raise ValueError("Must provide strings.")
+
+        super().__setitem__(key, value)
+
+    def fillna(self, value=None, method=None, limit=None):
+        # TODO: validate dtype
+        return super().fillna(value, method, limit)
+
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+        if isinstance(dtype, StringDtype):
+            if copy:
+                return self.copy()
+            return self
+        return super().astype(dtype, copy)
+
+    def __add__(self, other):
+        return _add(self, other)
+
+    def __radd__(self, other):
+        return _add(self, other, reversed=True)
+
+    def _reduce(self, name, skipna=True, **kwargs):
+        raise TypeError("Cannot perform reduction '{}' with string dtype".format(name))
+
+    def value_counts(self, dropna=False):
+        from pandas import value_counts
+
+        return value_counts(self._ndarray, dropna=dropna)
+
+
+def _add(array, other, reversed=False):
```

Comment:
`reversed` is a python keyword.  can you use e.g. is_reversed?

Reply:
alternatively, you could import radd from ops and pass either operator.add or ops.radd.  That might generalize more nicely for e.g. mul and rmul.